### PR TITLE
[FEAT] implement range operation and data streaming

### PR DESCRIPTION
**Pull request by andrewgazelka**

### CODE REVIEW

File: src/daft-connect/src/convert.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                let stream: Pin<Box<dyn Stream<Item = DaftResult<Table>>>> =
+                    Box::pin(stream::iter(tables));
+
+                stream
+            }
+            Err(err) => Box::pin(stream::once(async { Err(err) })),
+        });
+
+    Ok(stream)
+}
```
Comment (universalmind303) [inline src/daft-connect/src/convert.rs:45]: can we move this into `daft_local_execution`. I think this'll be helpful for my sql work. 

also nit: the name `run_local` is imo pretty bad _(I know the original one is also called `run_local`)_. Could we rename this to something more descriptive like `execute_logical_plan`

File: src/daft-connect/src/convert/expression.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        LiteralType::Date(input) => daft_dsl::LiteralValue::Date(input),
+        LiteralType::Decimal(_)
+        | LiteralType::Timestamp(_)
+        | LiteralType::TimestampNtz(_)
+        | LiteralType::CalendarInterval(_)
+        | LiteralType::YearMonthInterval(_)
+        | LiteralType::DayTimeInterval(_)
+        | LiteralType::Array(_)
+        | LiteralType::Map(_)
+        | LiteralType::Struct(_) => bail!("unimplemented"),
```
Comment (universalmind303) [inline src/daft-connect/src/convert/expression.rs:116]: i'd change this to match on the wildcard and return the type in the error message. 
```suggestion
        lit => bail!("Literal Value '{lit}' is not yet supported"),
```

File: src/daft-connect/src/convert/plan_conversion.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            }
+            RelType::Read(read) => parse_read(read),
+            _ => bail!("Unsupported relation type: {rel_type:?}"),
+        }
+    });
+
+    scope.join().unwrap()
+}
+
+fn parse_filter(filter: Filter) -> Result<LogicalPlanBuilder> {
```
Comment (universalmind303) [inline src/daft-connect/src/convert/plan_conversion.rs:35]: nit: i don't think were parsing the filter, but instead attempting to plan/convert it to our logical filter.

File: src/daft-connect/src/convert/plan_conversion.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        } = alias;
+
+        if name.len() != 1 {
+            bail!("Alias name must have exactly one element");
+        }
+        let name = name[0].as_str();
+
+        if metadata.is_some() {
+            bail!("Metadata is not yet supported");
+        }
```
Comment (universalmind303) [inline src/daft-connect/src/convert/plan_conversion.rs:70]: do we need to bail? can we just ignore the metadata instead.

File: src/daft-connect/src/convert/plan_conversion.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    for col_name in existing_columns {
+        new_exprs.push(daft_dsl::col(col_name));
+    }
+
+    input_plan
+        .select(new_exprs)
+        .wrap_err("selecting new expressions")
+}
+
+fn parse_read(read: Read) -> Result<LogicalPlanBuilder> {
```
Comment (universalmind303) [inline src/daft-connect/src/convert/plan_conversion.rs:94]: same comment as `parse_filter`

File: src/daft-connect/src/convert/plan_conversion.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+    let read_type = read_type.ok_or_else(|| eyre!("type is None"))?;
+
+    match read_type {
+        ReadType::NamedTable(_) => bail!("Named tables are not yet supported"),
+        ReadType::DataSource(data_source) => parse_data_source(data_source),
+    }
+}
+
+fn parse_data_source(data_source: DataSource) -> Result<LogicalPlanBuilder> {
```
Comment (universalmind303) [inline src/daft-connect/src/convert/plan_conversion.rs:110]: same comment as `parse_filter`

File: src/daft-connect/src/convert/plan_conversion.rs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    let condition = condition.ok_or_else(|| eyre!("condition is None"))?;
+    let condition =
+        expression::convert_expression(condition).wrap_err("converting to daft expression")?;
+    let condition = Arc::new(condition);
+
+    input_plan.filter(condition).wrap_err("applying filter")
+}
+
+fn parse_with_columns(with_columns: WithColumns) -> Result<LogicalPlanBuilder> {
```
Comment (universalmind303) [inline src/daft-connect/src/convert/plan_conversion.rs:48]: same comment as `parse_filter`

File: src/daft-connect/src/lib.rs [hunk 8]
```diff
@@ -172,9 +259,49 @@ impl SparkConnectService for DaftSparkConnectService {
...
+                    parsed: Some(result),
+                };
+
+                let response = AnalyzePlanResponse {
+                    session_id,
+                    server_side_session_id: String::new(),
+                    result: Some(analyze_plan_response::Result::DdlParse(schema)),
+                };
+
+                println!("response: {response:#?}");
```
Comment (universalmind303) [inline src/daft-connect/src/lib.rs:299]: remove println

File: src/daft-local-execution/src/run.rs [hunk 9]
```diff
@@ -92,6 +124,9 @@ impl NativeExecutor {
                 results_buffer_size,
             )
         })?;
+
+        let out = BlockingStreamIter::new(out);
```
Comment (universalmind303) [inline src/daft-local-execution/src/run.rs:128]: can you add a comment on why we need to block here?

File: src/daft-logical-plan/src/optimization/rules/eliminate_cross_join.rs
Comment (universalmind303) [inline src/daft-logical-plan/src/optimization/rules/eliminate_cross_join.rs:1]: can we remove these unrelated changes

File: src/daft-scan/src/hive.rs
Comment (universalmind303) [inline src/daft-scan/src/hive.rs:1]: can we remove these unrelated changes

File: src/daft-connect/src/session.rs [hunk 10]
```diff
@@ -8,6 +8,12 @@ pub struct Session {
     /// Also, <https://users.rust-lang.org/t/hashmap-vs-btreemap/13804/4>
     config_values: BTreeMap<String, String>,
 
+    #[expect(
+        unused,
+        reason = "this will be used in the future especially to pass spark connect tests"
+    )]
+    tables_by_name: HashMap<String, daft_table::Table>,
+
```
Comment (universalmind303) [inline src/daft-connect/src/session.rs:16]: couldn't we just track the logical plan instead of the `Table` and defer execution? Storing all of the materialized tables in a hashmap will have a **huge** memory footprint.

File: src/daft-connect/src/convert.rs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    let stream = daft_local_execution::run_local(&physical_plan, psets, cfg, None)?;
+
+    let stream = stream
+        .map(|partition| match partition {
+            Ok(partition) => partition.get_tables().map_err(DaftError::from),
+            Err(err) => Err(err),
+        })
+        .flat_map(|tables| match tables {
+            Ok(tables) => {
+                let tables = Arc::try_unwrap(tables).unwrap();
```
Comment (universalmind303) [inline src/daft-connect/src/convert.rs:33]:
```suggestion
                let tables = Arc::unwrap_or_clone(tables);
```

**Review by universalmind303 — COMMENTED** [review]
overall looks good to me, a few minor suggestions and comments.

File: src/daft-connect/src/session.rs (hunk 10 above)
Comment (andrewgazelka) [reply src/daft-connect/src/session.rs]: this is actually not relevant to this PR but should be discussed in the future  5aa8f7ab8041291c807bd5717378b2b953cee676

File: src/daft-logical-plan/src/optimization/rules/eliminate_cross_join.rs
Comment (andrewgazelka) [reply src/daft-logical-plan/src/optimization/rules/eliminate_cross_join.rs:1]: bfaac5b699a09c25d64997a87af19dce52cb36a6

File: src/daft-local-execution/src/run.rs (hunk 9 above)
Comment (andrewgazelka) [reply src/daft-local-execution/src/run.rs:128]: I am not completely sure but aren't we not leveraging python async really? → we are just blocking for a task to complete

File: src/daft-scan/src/hive.rs
Comment (andrewgazelka) [reply src/daft-scan/src/hive.rs:1]: done

File: src/daft-connect/src/convert.rs (hunk 1 above)
Comment (andrewgazelka) [reply src/daft-connect/src/convert.rs]: actually going to save this for another PR as it looks like this isn't actually used anywhere yet for ranges

File: src/daft-scan/src/hive.rs
Comment (universalmind303) [reply src/daft-scan/src/hive.rs:1]: Looks like they're still in the PR

**Review by universalmind303 — APPROVED** [review]
This looks mostly fine to me. I think some function/struct names could be clearer to better express their purpose - but that's just my opinion.

**Comment by andrewgazelka** [comment]
> This looks mostly fine to me. I think some function/struct names could be clearer to better express their purpose - but that's just my opinion.

will do in future PRs
